### PR TITLE
Add support for configuration discovery

### PR DESCRIPTION
**Pull request by ashutosh-narkar**
Signed-off-by: Ashutosh Narkar <anarkar4387@gmail.com>

### CODE REVIEW

File: runtime/runtime.go [hunk 1]
```diff
@@ -590,6 +615,118 @@ func loadConfig(id string, store storage.Store, configFile string) (*loadedConfi
...
+	)
+
+	rs, err := rego.Eval(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var newConfig []byte
+	if len(rs) == 0 {
+		return newConfig, nil
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply runtime/runtime.go]: done

File: runtime/runtime.go [hunk 2]
```diff
@@ -590,6 +615,118 @@ func loadConfig(id string, store storage.Store, configFile string) (*loadedConfi
...
+
+	query := "data"
+	if discoveryConfig.Path != "" {
+		query = fmt.Sprintf("data%v", strings.Replace(discoveryConfig.Path, "/", ".", -1))
+	}
+
+	return processBundle(ctx, b, query)
+}
+
+func processBundle(ctx context.Context, bundle bundleApi.Bundle, query string) ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply runtime/runtime.go]: Done

File: runtime/runtime.go [hunk 3]
```diff
@@ -590,6 +615,118 @@ func loadConfig(id string, store storage.Store, configFile string) (*loadedConfi
...
+	}
+
+	if config.Discovery == nil {
+		return false
+	}
+	return true
+}
+
+func discoveryHandler(ctx context.Context, bs []byte, manager *plugins.Manager) ([]byte, error) {
+	var config struct {
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply runtime/runtime.go]: Avoided

File: runtime/runtime.go [hunk 4]
```diff
@@ -590,6 +615,118 @@ func loadConfig(id string, store storage.Store, configFile string) (*loadedConfi
 	return c, nil
 }
 
+func isDiscoveryEnabled(bs []byte) bool {
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply runtime/runtime.go]: Created new `discovery` package

File: docs/book/configuration.md [hunk 5]
```diff
@@ -86,4 +86,60 @@ multiple services.
 | `decision_logs.reporting.buffer_size_limit_bytes` | `int64` | No | Decision log buffer size limit in bytes. OPA will drop old events from the log if this limit is exceeded. By default, no limit is set. |
 | `decision_logs.reporting.upload_size_limit_bytes` | `int64` | No (default: `32768`) | Decision log upload size limit in bytes. OPA will chunk uploads to cap message body to this limit. |
 | `decision_logs.reporting.min_delay_seconds` | `int64` | No (default: `300`) | Minimum amount of time to wait between uploads. |
-| `decision_logs.reporting.max_delay_seconds` | `int64` | No (default: `600`) | Maximum amount of time to wait between uploads. |
\ No newline at end of file
+| `decision_logs.reporting.max_delay_seconds` | `int64` | No (default: `600`) | Maximum amount of time to wait between uploads. |
+
+## Discovery
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply docs/book/configuration.md]: Created a new `Discovery` section

File: docs/book/configuration.md [hunk 6]
```diff
@@ -86,4 +86,60 @@ multiple services.
...
+
+OPA's configuration files contain the configuration for management functionality such as `bundles`, `status`, `decision logs` . As OPA gains more management functionality, it's configuration would have to be extended and could get difficult to manage. To simplify deployment of OPA and enable better centralized management, OPA can discover it's configuration from a remote endpoint.
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply docs/book/configuration.md]: Updated the doc

**Comment by tsandall** [comment]
I did a bit more manual testing with the latest version:

```
$ opa version
Version: 0.10.2-dev
Build Commit: df45db42
Build Timestamp: 2018-11-22T19:52:16Z
Build Hostname: blackbox
```

My notes are below. I think there will be a little bit more churn on the interfaces and implementation based on the last round of testing so I think it's worth addressing some of these things before diving into the code itself.

## Notes

- Plugins are being reconfigured even when configuration has not changed. You
  can see this happening in the logs below. No change was made in the 5 seconds
  between boot and when the reconfiguration occurred.

    ```
    $ cat config.yaml
    services:
      test: <REDACTED>

    discovery:
      service: test
      path: torin/disco/disco
      polling:
        min_delay_seconds: 5
        max_delay_seconds: 5

    $ opa run -c config.yaml -s
    INFO[2018-11-22T11:58:42-08:00] First line of log stream.                     addrs="[:8181]" insecure_addr=
    INFO[2018-11-22T11:58:42-08:00] New configuration successfully downloaded. Now updating plugins.  config=discovery
    INFO[2018-11-22T11:58:42-08:00] Log upload skipped.                           plugin=decision_logs
    INFO[2018-11-22T11:58:42-08:00] Bundle downloaded and activated successfully. Etag updated to "84bebb11107eebf6ac0aeff0831197c8fa94d04d32d23ed229e8e41e3d18ce60".  name=torin/apiauthz plugin=bundle
    INFO[2018-11-22T11:58:42-08:00] Status update sent successfully.              plugin=status
    INFO[2018-11-22T11:58:47-08:00] New configuration successfully downloaded. Now updating plugins.  config=discovery
    INFO[2018-11-22T11:58:47-08:00] Decision Logs plugin reconfigured successfully.  plugin=decision_logs
    INFO[2018-11-22T11:58:47-08:00] Log upload skipped.                           plugin=decision_logs
    INFO[2018-11-22T11:58:47-08:00] Bundle plugin reconfigured successfully.      name=torin/apiauthz plugin=bundle
    INFO[2018-11-22T11:58:47-08:00] Status plugin reconfigured successfully.      plugin=status
    INFO[2018-11-22T11:58:47-08:00] Status update sent successfully.              plugin=status
    ```

- Plugin reconfiguration errors leave the plugin in a broken state which is
  only explicitly reported in the logs. We need to rethink how this is handled.
  Ideally the new configuration would not be activated unless all of the
  plugins accepted it. If there is an error in the configuration for any of the
  plugins, those errors should be reported in the logs and in the status API.
  We need to make the process 2-phase. First provide the new configuration to
  each plugin and ask the plugin to validate it. If any of the plugins return a
  validation error, reject the new configuration. If they all accept the
  configuration, they can be reconfigured. Configuration errors should be
  reported via the status plugin if it's active.

- The `default_decision` setting is not set by the discovered configuration.
  The same is probably happening for the `default_authorization_decision`
  setting (haven't checked the code.)

- Nit: Discovery plugin should log the configuration changes in a
  human-readable format. Listing the set of plugins that have been reconfigured
  is sufficient (at INFO level).

- Nit: Configuration errors reported as discovery errors during boot. For
  example, the configuration below contains a typo in the bundle polling
  configuration.

    ```
    $ cat config.yaml
    services:
      test: <REDACTED>
    bundle:
      service: test
      polling:
        min_delay_seonds: 100
        max_delay_seconds: 100
    #discovery:
    #  service: test
    #  path: torin/disco/disco
    #  polling:
    #    min_delay_seconds: 5
    #    max_delay_seconds: 5

    $ opa run -c config.yaml -s
    error: discovery: invalid bundle name ""
    ```

- Nit: Use `name` instead of `path` in discovery configuration to be consistent
  with the bundle configuration.

**Comment by JAORMX** [comment]
>     * Plugins are being reconfigured even when configuration has not changed. You
>       can see this happening in the logs below. No change was made in the 5 seconds
>       between boot and when the reconfiguration occurred.

The only thing I can think of to address this is to require appropriate HTTP headers from the server that provides the configuration.

Given that it's likely that the configuration will be cached on the server side, the ETag could be checked. We could then store the ETag and compare it with subsequent calls to the server. If the server would support this, it would also save quite a bit of bandwith, since the response is not needed if ETag is supported, as that indicates that the content hasn't changed.

Alternatively, the server could answer with a Last-Modified header, which could be stored in OPA for comparison purposes. It could also issue an "Expires" header, which would tell OPA if the configuration needs to be re-fetched.

**Comment by tsandall** [comment]
@JAORMX thanks for the input! In this case, we were checking whether the configuration generated from the bundle had changed--the check as just broken. Using the ETag would make sense although currently we're just checking the configuration object for equality. Either approach should work. I'd rather avoid time-based approaches.

**Comment by ashutosh-narkar** [comment]
> I did a bit more manual testing with the latest version:
> 
> ```
> $ opa version
> Version: 0.10.2-dev
> Build Commit: df45db42
> Build Timestamp: 2018-11-22T19:52:16Z
> Build Hostname: blackbox
> ```
> My notes are below. I think there will be a little bit more churn on the interfaces and implementation based on the last round of testing so I think it's worth addressing some of these things before diving into the code itself.
> 
> ## Notes
> * Plugins are being reconfigured even when configuration has not changed. You
>   can see this happening in the logs below. No change was made in the 5 seconds
>   between boot and when the reconfiguration occurred.
>   ```
>   $ cat config.yaml
>   services:
>     test: <REDACTED>
>   
>   discovery:
>     service: test
>     path: torin/disco/disco
>     polling:
>       min_delay_seconds: 5
>       max_delay_seconds: 5
>   
>   $ opa run -c config.yaml -s
>   INFO[2018-11-22T11:58:42-08:00] First line of log stream.                     addrs="[:8181]" insecure_addr=
>   INFO[2018-11-22T11:58:42-08:00] New configuration successfully downloaded. Now updating plugins.  config=discovery
>   INFO[2018-11-22T11:58:42-08:00] Log upload skipped.                           plugin=decision_logs
>   INFO[2018-11-22T11:58:42-08:00] Bundle downloaded and activated successfully. Etag updated to "84bebb11107eebf6ac0aeff0831197c8fa94d04d32d23ed229e8e41e3d18ce60".  name=torin/apiauthz plugin=bundle
>   INFO[2018-11-22T11:58:42-08:00] Status update sent successfully.              plugin=status
>   INFO[2018-11-22T11:58:47-08:00] New configuration successfully downloaded. Now updating plugins.  config=discovery
>   INFO[2018-11-22T11:58:47-08:00] Decision Logs plugin reconfigured successfully.  plugin=decision_logs
>   INFO[2018-11-22T11:58:47-08:00] Log upload skipped.                           plugin=decision_logs
>   INFO[2018-11-22T11:58:47-08:00] Bundle plugin reconfigured successfully.      name=torin/apiauthz plugin=bundle
>   INFO[2018-11-22T11:58:47-08:00] Status plugin reconfigured successfully.      plugin=status
>   INFO[2018-11-22T11:58:47-08:00] Status update sent successfully.              plugin=status
>   ```
> * Plugin reconfiguration errors leave the plugin in a broken state which is
>   only explicitly reported in the logs. We need to rethink how this is handled.
>   Ideally the new configuration would not be activated unless all of the
>   plugins accepted it. If there is an error in the configuration for any of the
>   plugins, those errors should be reported in the logs and in the status API.
>   We need to make the process 2-phase. First provide the new configuration to
>   each plugin and ask the plugin to validate it. If any of the plugins return a
>   validation error, reject the new configuration. If they all accept the
>   configuration, they can be reconfigured. Configuration errors should be
>   reported via the status plugin if it's active.
> * The `default_decision` setting is not set by the discovered configuration.
>   The same is probably happening for the `default_authorization_decision`
>   setting (haven't checked the code.)
> * Nit: Discovery plugin should log the configuration changes in a
>   human-readable format. Listing the set of plugins that have been reconfigured
>   is sufficient (at INFO level).
> * Nit: Configuration errors reported as discovery errors during boot. For
>   example, the configuration below contains a typo in the bundle polling
>   configuration.
>   ```
>   $ cat config.yaml
>   services:
>     test: <REDACTED>
>   bundle:
>     service: test
>     polling:
>       min_delay_seonds: 100
>       max_delay_seconds: 100
>   #discovery:
>   #  service: test
>   #  path: torin/disco/disco
>   #  polling:
>   #    min_delay_seconds: 5
>   #    max_delay_seconds: 5
>   
>   $ opa run -c config.yaml -s
>   error: discovery: invalid bundle name ""
>   ```
> * Nit: Use `name` instead of `path` in discovery configuration to be consistent
>   with the bundle configuration.

Following changes pushed:
* Plugins not reconfigured when config unchanged.

* Plugin configuration split into 2 steps: Validation and Configuration. If validation fails, error logged and reported via status plugin if active. If validation succeeds, plugins configured/re-configured.

* Currently `default_decision` and `default_authorization_decision` set only on boot up. 

* Plugins that are configured/re-configured are logged at INFO level.

* Updated how configuration errors reported during boot.

* Used `name` instead of `path` in discovery configuration

File: discovery/discovery_test.go [hunk 7]
```diff
@@ -0,0 +1,483 @@
...
+
+func TestConfigDiscoveryDisabled(t *testing.T) {
+	_, result := isDiscoveryEnabled([]byte{})
+
+	if result {
+		t.Fatal("Expected discovery not to be enabled")
+	}
+}
+
+func TestGetDiscoveryServicePath(t *testing.T) {
```
Comment (tsandall) [inline discovery/discovery_test.go]: This test should be table-driven.

File: docs/book/SUMMARY.md [hunk 8]
```diff
@@ -118,6 +119,11 @@
   * [Service API](decision_logs.md#decision-log-service-api)
 * [Monitoring](monitoring.md)
   * [Prometheus](monitoring.md#prometheus)
+* [Discovery](discovery.md)
+  * [Service API](discovery.md#discovery-service-api)
+* [Monitoring & Diagnostics](monitoring-diagnostics.md)
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply docs/book/SUMMARY.md]: Removed.

File: discovery/discovery.go [hunk 9]
```diff
@@ -0,0 +1,845 @@
...
+	"github.com/open-policy-agent/opa/storage"
+	"github.com/open-policy-agent/opa/storage/inmem"
+	"github.com/open-policy-agent/opa/util"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+type bundlePluginListener string
+
+// Discovery contains the result of configuration discovery.
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply discovery/discovery.go]: Updated the comment.

File: discovery/discovery.go [hunk 10]
```diff
@@ -0,0 +1,845 @@
...
+	// min amount of time to wait following a failure
+	minRetryDelay          = time.Millisecond * 100
+	defaultMinDelaySeconds = int64(60)
+	defaultMaxDelaySeconds = int64(120)
+
+	errCode         = "discovery_error"
+	discoveryPlugin = "discovery"
+)
+
+// TODO(tsandall): revisit how plugins are wired up to the manager and how
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply discovery/discovery.go]: Removed.

File: discovery/discovery.go [hunk 11]
```diff
@@ -0,0 +1,845 @@
...
+// everything is started and stopped. We could introduce a package-scoped
+// plugin registry that allows for (dynamic) init-time plugin registration.
+
+// New takes the provided configuration and if needed fetches
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply discovery/discovery.go]: Updated comment.

File: discovery/discovery.go [hunk 12]
```diff
@@ -0,0 +1,845 @@
...
+// OPA's new configuration from a remote server. It then instantiates the
+// plugins using either the provided configuration or the one fetched from
+// the remote server.
+func New(ctx context.Context, opaID string, store storage.Store, configFile string, registeredPlugins map[string]plugins.PluginInitFunc) (*Discovery, error) {
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply discovery/discovery.go]: Added a new `Params` struct to hold these params.

File: discovery/discovery.go [hunk 13]
```diff
@@ -0,0 +1,845 @@
...
+
+	config, discoveryEnabled = isDiscoveryEnabled(bs)
+	if discoveryEnabled {
+		discPathConfig, err = validateAndInjectDefaults(config, m.Services())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	p, err := initPlugins(m, bs, registeredPlugins)
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply discovery/discovery.go]: If discovery is enabled and any other plugin is defined in the bootstrap configuration, the following error `plugins cannot be specified in the bootstrap configuration when discovery enabled` will be returned.

File: discovery/discovery.go [hunk 14]
```diff
@@ -0,0 +1,845 @@
...
+	minSeconds := int64(time.Duration(min) * time.Second)
+	discoveryConfig.Polling.MinDelaySeconds = &minSeconds
+
+	maxSeconds := int64(time.Duration(max) * time.Second)
+	discoveryConfig.Polling.MaxDelaySeconds = &maxSeconds
+
+	return &discoveryConfig, nil
+}
+
+func configureDefaultDecision(bs []byte) (ast.Ref, ast.Ref, error) {
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply discovery/discovery.go]: Changed the method name.

File: discovery/discovery.go [hunk 15]
```diff
@@ -0,0 +1,845 @@
...
+type bundlePluginListener string
+
+// Discovery contains the result of configuration discovery.
+type Discovery struct {
+	Manager                      *plugins.Manager
+	Plugins                      map[string]plugins.Plugin
+	DefaultDecision              ast.Ref
+	DefaultAuthorizationDecision ast.Ref
+
+	discoveryEnabled    bool
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply discovery/discovery.go]: `discoveryEnabled ` changed to `dynamic `
`discoveryPathConfig` changed to `config`

File: discovery/discovery.go [hunk 16]
```diff
@@ -0,0 +1,845 @@
...
+
+type discoveryConfig struct {
+	Discovery json.RawMessage `json:"discovery"`
+}
+
+type discoveryPathConfig struct {
+	Name        *string       `json:"name"`
+	Prefix      *string       `json:"prefix"`
+	Polling     PollingConfig `json:"polling"`
+	serviceURL  string
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply discovery/discovery.go]: Done

File: discovery/discovery.go [hunk 17]
```diff
@@ -0,0 +1,845 @@
...
+	serviceHost string
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply discovery/discovery.go]: Done.

File: discovery/discovery.go [hunk 18]
```diff
@@ -0,0 +1,827 @@
...
+	}
+
+	compiler := ast.NewCompiler()
+	if compiler.Compile(modules); compiler.Failed() {
+		return nil, false, compiler.Errors
+	}
+
+	store := inmem.NewFromObject(b.Data)
+
+	rego := rego.New(
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply discovery/discovery.go]: Added `rego.Runtime`

**Comment by tsandall** [comment]
This is at the point where people can start playing with it. I'm going to squash and merge these changes. We can address the remaining code quality comments in a subsequent PR.

**Pull request merged**